pallets/randomness/src/lib.rs

```diff
@@ -0,0 +1,433 @@
+			earliest_epoch: u64,
+		},
+		RandomnessRequestedLocal {
+			id: RequestId,
+			refund_address: H160,
+			contract_address: H160,
+			fee: BalanceOf<T>,
+			gas_limit: u64,
+			salt: H256,
+			earliest_block: T::BlockNumber,
+		},
+		RequestFulfilled {
+			id: RequestId,
+		},
+		RequestFeeIncreased {
+			id: RequestId,
+			new_fee: BalanceOf<T>,
+		},
+		RequestExpirationExecuted {
+			id: RequestId,
+		},
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn requests)]
+	/// Randomness requests not yet fulfilled or purged
+	pub type Requests<T: Config> = StorageMap<_, Twox64Concat, RequestId, RequestState<T>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn request_count)]
+	/// Number of randomness requests made so far, used to generate the next request's uid
+	pub type RequestCount<T: Config> = StorageValue<_, RequestId, ValueQuery>;
+
+	/// Current local per-block VRF randomness
+	/// Set in `on_initialize`, before it will contain the randomness for this block
+	#[pallet::storage]
+	#[pallet::getter(fn local_vrf_output)]
+	pub type LocalVrfOutput<T: Config> = StorageValue<_, Option<T::Hash>, ValueQuery>;
+
+	/// Current VRF input
+	/// Set in `on_finalize` of last block
+	/// Used in `on_initialize` of current block to validate VRF output
+	#[pallet::storage]
+	#[pallet::getter(fn current_vrf_input)]
+	pub(crate) type CurrentVrfInput<T: Config> = StorageValue<_, VrfInput<Slot, T::Hash>>;
+
+	/// Relay time information to determine when to update randomness results based on when
+	/// epoch and relay block change
+	#[pallet::storage]
+	#[pallet::getter(fn relay_time)]
+	pub(crate) type RelayTime<T: Config> =
+		StorageValue<_, RelayTimeInfo<T::BlockNumber, u64>, ValueQuery>;
+
+	/// Ensures the mandatory inherent was included in the block
+	#[pallet::storage]
+	#[pallet::getter(fn inherent_included)]
+	pub(crate) type InherentIncluded<T: Config> = StorageValue<_, ()>;
+
+	/// Records whether this is the first block (genesis or runtime upgrade)
+	#[pallet::storage]
+	#[pallet::getter(fn not_first_block)]
+	pub type NotFirstBlock<T: Config> = StorageValue<_, ()>;
+
+	/// Snapshot of randomness to fulfill all requests that are for the same raw randomness
+	/// Removed once $value.request_count == 0
+	#[pallet::storage]
+	#[pallet::getter(fn randomness_results)]
+	pub(crate) type RandomnessResults<T: Config> =
+		StorageMap<_, Twox64Concat, RequestType<T>, RandomnessResult<T::Hash>>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Populates the `RandomnessResults` that are due this block with the raw values
+		/// This inherent is a workaround to run code in every block after
+		/// `ParachainSystem::set_validation_data` but before all extrinsics.
+		/// the relevant storage item to validate the VRF output in this pallet's `on_initialize`
+		// This should go into on_post_inherents when it is ready
+		// https://github.com/paritytech/substrate/pull/10128
+		// TODO: weight
+		#[pallet::weight((10_000, DispatchClass::Mandatory))]
```

Comment:
We should write a test case that makes sure this fn is always `Mandatory`. It's too easy to accidentally remove this when updating the weight...

Reply:
There is a check in `on_finalize` that ensures this inherent is always included. The relevant storage item is `InherentIncluded`

Reply:
Yes, but we don't want to wait until we hit this on-chain.

By the way, this exact problem has occurred before ;) In that case, it caused a lot of headache to find the issue (which stalled the chain), fix it, create a whole new release, etc. Having a unit test would catch this immediately if it were a problem.

Reply:
Actually, it's even more nuanced than that. We got lucky we hit it when we did.

If it becomes not `Mandatory` by mistake, it will still be included except in rare cases where we go overweight during initialization.

See https://github.com/PureStake/nimbus/pull/60 for more details

Reply:
Added this in 2d727a5